### ISSUE TITLE
is clone supposed to work with references?

### ISSUE THREAD

**Issue by jole78** [mobxjs/mobx-state-tree#773]
Hi,

I have a model that looks like this:
```
const Asset = types.model("Asset", {
    uuid: types.identifier(),
    name: types.string,
    type: types.maybe(types.reference(AssetType)),
    owner: types.string,
});
...// in aonther file...
const AssetType = types.model("AssetType", {
    uuid: types.identifier(),
    name: types.string,
});
```
where, as you can see the asset has a type which can be an asset type. Everything works great when working in read-only mode...like listing assets and display the asset.type.name etc...

However, I'm basically "following" along with the code from @mweststrate egghead.io course on mst (which is a great resource btw) https://github.com/mweststrate/mst-course and the "advice" seems to be to clone the item you're editing...so basically I'm doing this in an event handler:

```
    onToggleEdit = (event) => {
        this.setState({
            isEditing: true,
            asset: clone(this.props.asset),
        });
    };  
```
The asset type is the supposed to be selected from a "picker" component (a dropdown basically).However, when I try to forward the asset.type (from the clone...e.g. this.state.asset.type) to the picker (as the value prop) I'm getting this thrown by mst:
`[mobx-state-tree] Failed to resolve reference '0c0fede0-4244-11e8-af22-d19eb1215781' to type 'AssetType' (from node: /type)`

I tried to console.log the cloned asset to access the type and the same happens there. I also tried the same with the code from (https://github.com/mweststrate/mst-course) and the same happens there...like this in app.js

```
    onSelectUser = event => {
        this.setState({ selectedUser: event.target.value })
        const { group } = this.props
        const selectedUser = group.users.get(event.target.value)
        console.log(clone(selectedUser))
    }
```

`[Exception: Error: [mobx-state-tree] Failed to resolve reference of type late(function () { return User; }): 'ba32' (in: /recipient)`

so, is this not intended to work or did I find a bug or am I totally misunderstanding this...?? 

any help is appreciated :)

**Comment by mweststrate** [comment]
References can only be resolved if the object is part of the same tree, so
you need to either make clone part of the tree or use a custom reference
type

Op di 17 apr. 2018 18:58 schreef Johan Leino <notifications@github.com>:

**Comment by jole78** [comment]
thanks for your reply @mweststrate. I'm not quite there yet with tree semantics I think ;)

Given this:

```
const AssetState = types.model({
    types: types.optional(types.array(AssetType), []),
    items: types.optional(types.array(Asset), []),
});
```
Are they not part of the same tree?? ...or what does that mean?

**Comment by mweststrate** [comment]
They are part of the same tree, but a clone of them isn't (it is just a
completely new value), until you put it somewhere in the tree :). Part of
the same tree is about the instances, not the types. IN other words, is
there a way in which you could walk from one object to another, by either
accessing properties or using the parent? So clone(x.y.z[0]) is not part of
the tree of x, but x.y.z.push(clone(x.y.z[0])) is.

Op di 17 apr. 2018 om 21:32 schreef Johan Leino <notifications@github.com>:

**Comment by jole78** [comment]
Ah...now I see. So when I clone an object that has references they are "lost" sort of. 
I actually decided to go with a different solution for now at least.
Thanks anyway!

**Comment by awel-light** [comment]
Hi, I went down the same road as OP.. My model has many references which renders clone useless. I wonder what the best next step would be? Put the clones somewhere in the tree, or manual cloning, operating on snapshots..? What path did you @jole78 take?  
Which begs the question, what about having some mechanism such that the clone is still attached to the tree so the cloning just works?

**Comment by jole78** [comment]
@awel-light don't know if it will help you since this was quite a while ago but...

My original use case was to do an edit form for an asset basically.

What I did was something like this component (showing only relevant code) :

```javascript
@observer
class AssetForm extends React.Component {
    state = {
        asset: this.getInitialState(),
    };

    getInitialState(props = this.props) {
        if (props.editmode && props.asset) {
            return edit(props.asset);
        } else {
            // createInstance just creates a new asset for me
            return createInstance();
        }
    }

    onSubmit = (event) => {
        const { editmode } = this.props;
        if (editmode) {
            // the essence is here where I'm applying the snapshot of the edited asset
            applySnapshot(this.props.asset, getSnapshot(this.state.asset));
        } else {
            this.props.assets.push(this.state.asset);
            this.reset();
        }
    };


    render() {
        const { asset } = this.state;
        const { editmode } = this.props;

        return (
            // FORM CODE GOES HERE
        );
    }
}

...edit function...
export function edit(node, modifiable = true) {
    const instance = clone(node);
    if (modifiable) {
        unprotect(instance);
    }

    return instance;
}
```

**Comment by mattruby** [comment]
I've run into the same issue with references.  I like the flexibility, however, I do wish there was a way to limit refs to certain branches of a tree.

**Comment by awel-light** [comment]
@jole78 I don't see how your code solves the reference-cloning issue?

I ended up, for now, to get a snapshot of the object and edit that (using immer), then apply the snapshot back to the model on submit. This works fine, and I really like immer, however this is bypassing MST completely.

I'll also try to create a new top model and put the clones in a branch for UI-related state; even though I'd rather keep this volatile state out of the tree, it's probably better than bypassing MST.

@mweststrate thoughts on this one?

**Comment by jole78** [comment]
@awel-light you just asked me what path I took...didn't actually say I solved anything...apart from that I got it working in what I think was a suitable way. Still love MST! 😉

**Comment by xaviergonz** [comment]
you could also use applySnapshot over the original node with the getSnapshot from the clone once editing is done

**Comment by awel-light** [comment]
@xaviergonz cloning doesn't work because of the reference being disattached from the tree as in the original post :(
@jole78 yes thanks :) I am just trying to understand what is happening, still learning, and I didn't see how it could work.. Do you see the changes reflected in the whole tree while editing?
